woocommerce/google-listings-and-ads
# AppInputPriceControl component to parse and format price value

### Changes proposed in this Pull Request:

Closes #402 .

This PR adds a new `<AppInputPriceControl>` component and uses it in the following places:

- Setup Ads Step 2: Budget section.
- Setup MC Step 3: Shipping rate section
    - Shipping rate inline edit
    - Edit shipping rate modal
    - Add shipping rate modal
    - The "I offer free shipping for products priced over" inline edit. 

The `<AppInputPriceControl>` adds a few main functionalities over the normal `<AppInputControl>`: 

- Accepts `value` of number type and formats it based on the store's currency settings.
- During `onChange` event, it parses user input string value into number value and propagate it up to the parent component via `onChange`.
- During `onBlur` event, it would again parse the string value seen by the user and propagate it up to the parent component via `onChange`.
    - This is to make prevent situation where form state has a number value `12345.6789` but the string value seen by the user has been formatted as `"12345.68"`, an inconsistency between data and user expectation.
- Display the store's currency code in the input control's suffix.

### Screenshots:

Demo video with my voice: https://d.pr/v/npFNTU

The following images show the AppInputPriceControl with the following store currency settings:

- Currency: Polish
- Thousand separator: `.`
- Decimal separator: `,`
- Number of decimals: `2`

![image](https://user-images.githubusercontent.com/417342/113323182-3f11ed80-9348-11eb-9c9d-926dc7776a36.png)

![image](https://user-images.githubusercontent.com/417342/113323452-8ac49700-9348-11eb-8577-030d7bbc9304.png)

### Detailed test instructions:

1. Go to https://gla1.test/wp-admin/admin.php?page=wc-settings and take note of the currency options settings.
2. Go to https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-ads
3. Proceed to Step 2. 
4. In the daily average cost and the monthly max fields, you should see the number `0` formatted based on your store currency settings. It should also have the store currency suffix. 
5. Enter value into the daily average cost. As you type, the monthly max field should be automatically updated and formatted based on the store currency settings.
    - The daily average cost value can contain thousand separator and decimal separator.
6. Enter invalid value (e.g. the string `"abcd"`) into the daily average cost. The monthly max would show 0 value. When the daily average cost field loses focus, it will be reset to 0.
7. Repeat the above steps but with different currency options settings.

### Changelog Note:

Use AppInputPriceControl to display and handle price values.
